File: algorithms/2_medium/C/01922_count_good_numbers/solution.cpp
constexpr long long MOD = 1'000'000'007;

class Solution {
    inline static long long pow_by_mod(long long base, long long expo, const long long mod_by) {
        long long res = 1;
        while (expo > 0) {
            if ((expo & 1) == 1) {
                res = (res * base) % mod_by;
            }
            base = (base * base) % mod_by;
            expo >>= 1;
        }
        return res;
    }
public:
    static int countGoodNumbers(const long long n) {
        if (n <= 0) { return -1; }

        const long long half = n / 2;

        const long long odd = pow_by_mod(4, half, MOD);
        const long long even = pow_by_mod(5, n - half, MOD);
        const long long res = (odd * even) % MOD;

        return static_cast<int>(res);
    }
};

File: algorithms/2_medium/C/01922_count_good_numbers/solution.rs
impl Solution {
    const MOD: i64 = 1_000_000_007;

    #[inline(always)]
    fn pow_by_mod(mut base: i64, mut expo: i64, mod_by: i64) -> i64 {
        let mut res: i64 = 1;
        while expo > 0 {
            if (expo & 1) == 1 {
                res = (res * base) % mod_by;
            }
            base = (base * base) % mod_by;
            expo >>= 1;
        }
        res
    }

    pub fn count_good_numbers(n: i64) -> i32 {
        assert!(n > 0);

        let half = n / 2;

        let odd = Self::pow_by_mod(4, half, Self::MOD);
        let even = Self::pow_by_mod(5, n - half, Self::MOD);
        let res = (odd * even) % Self::MOD;

        res as i32
    }
}
